eclipse/che
# Propagate plugin/editor component environment variables

### What does this PR do?
It introduces a new feature: propagating plugin/editor component environment variables.

This is archived by the following architecture change:
Now we propagate the whole Devfile along with WorkspaceConfig and InternalEnvironment and it makes it possible to access components configuration and apply it on plugin brokering phase.

What can be done better maybe in scope of separate PRs:
1. Now `EditorComponentToWorkspaceConfigApplier`/`PluginComponentToWorkspaceConfigApplier` modifies components that was originally passed. It's needed to make it possible to match component (configured with reference) with ChePlugin provided by plugin broker. As an improvement, we can get rid of it but Plugin Broker should provide additional information where it downloaded ChePlugin configuration. More see https://github.com/eclipse/che/pull/15435#discussion_r356049993
2. The whole Devfile is passed to WorkspaceConfig/InternalEnvironment and by the time being only Plugins/Editors components are used. As an alternative way we could propagate only used parts and there is no clear and easy way to do it smoothly. See more https://github.com/eclipse/che/pull/15435#discussion_r355943668

### How to test this PR
0. Build this branch or use the following image: `sleshchenko/che-server:env-vars`
1. Create a workspace with the following Devfile
<details>

<summary>Devfile</summary>

```yaml
metadata:
  name: env-override
components:
  - type: kubernetes
    reference: >-
      https://gist.githubusercontent.com/sleshchenko/86d28a2c4426bdd0cef57bdfe3c7a829/raw/555b20c10ff7d6519b80d0841a30735a2f4f2118/deployment.yaml
    alias: env-test
    env:
      - value: Serhii
        name: NAME
  - reference: https://gist.githubusercontent.com/sleshchenko/5e1118c93a17e6b7d4efc4cd6f88b164/raw/471553b59cf72510289b497e0f7a6f1c86e48450/meta.yaml
    memoryLimit: 1G
    type: cheEditor
    alias: editor
    env:
      - value: Hello My Dear Friend
        name: TEST
  - id: ms-vscode/go/latest
    type: chePlugin
    alias: plugin
    memoryLimit: 256M
    env:
      - value: Hello My Dear Friend
        name: TEST
apiVersion: 1.0.0
commands:
  - name: testk8s
    actions:
      - type: exec
        command: cd /hello && ls
        component: env-test
  - name: test-editor
    actions:
      - type: exec
        command: 'echo ${TEST}'
        component: editor
  - name: test-plugin
    actions:
      - type: exec
        command: 'echo ${TEST}'
        component: plugin
```

</details>
2. Run every command and make sure that output has values provided with env vars.
3. Check workspace pod deployment (kubectl or K8s/OS Dashboard) and make sure that memoryLimit is right.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/14343

#### Release Notes
N/A


#### Docs PR
https://github.com/eclipse/che-docs/pull/966
